Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean

```diff
@@ -38,504 +61,311 @@ universe u
-    · rw [IsOpenImmersion.isoOfRangeEq_hom_fac_assoc]
-      exact Subtype.range_coe
-
-end AlgebraicGeometry
+    simp_rw [sourceAffineLocally_morphismRestrict] at hs'
+    have := hs' r ⟨X.basicOpen (f.appLE ⊤ U le_top r.1), U.2.basicOpen (f.appLE ⊤ U le_top r.1)⟩
+      (by simpa [Scheme.Hom.appLE] using Scheme.basicOpen_restrict _ _ _)
+    rwa [IsAffineOpen.appLE_eq_away_map f (isAffineOpen_top Y) U.2,
+      ← h₁.is_localization_away_iff] at this
+
+end affineLocally
+
+/--
+`HasRingHomProperty P Q` is a type class asserting that `P` is local at the target and the source,
+and for `f : Spec B ⟶ Spec A`, it is equivalent to the ring hom property `Q`.
+To make the proofs easier, we state it instead as
+1. `Q` is local (See `RingHom.PropertyIsLocal`)
+2. `P f` if and only if `Q` holds for every `Γ(Y, U) ⟶ Γ(X, V)` for all affine `U`, `V`.
+-/
+class HasRingHomProperty (P : MorphismProperty Scheme.{u})
+    (Q : outParam (∀ {R S : Type u} [CommRing R] [CommRing S], (R →+* S) → Prop)) : Prop where
+  isLocal_ringHomProperty : RingHom.PropertyIsLocal Q
+  eq_affineLocally' : P = affineLocally Q
+
+namespace HasRingHomProperty
+
+variable (P : MorphismProperty Scheme.{u}) {Q} [HasRingHomProperty P Q]
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+lemma eq_affineLocally : P = affineLocally Q := eq_affineLocally'
+
+instance : HasAffineProperty P (sourceAffineLocally Q) where
+  isLocal_affineProperty := sourceAffineLocally_isLocal _
+    (HasRingHomProperty.isLocal_ringHomProperty P).respectsIso
+    (HasRingHomProperty.isLocal_ringHomProperty P).LocalizationPreserves
+    (HasRingHomProperty.isLocal_ringHomProperty P).ofLocalizationSpan
+  eq_targetAffineLocally' := eq_affineLocally P
+
+theorem appLE (H : P f) (U : Y.affineOpens) (V : X.affineOpens) (e) : Q (f.appLE U V e) := by
+  rw [eq_affineLocally P, affineLocally_iff_affineOpens_le] at H
+  exact H _ _ _
+
+theorem app_top (H : P f) [IsAffine X] [IsAffine Y] : Q (f.app ⊤) := by
+  rw [Scheme.Hom.app_eq_appLE]
+  exact appLE P f H ⟨_, isAffineOpen_top _⟩ ⟨_, isAffineOpen_top _⟩ _
+
+theorem comp_of_isOpenImmersion [IsOpenImmersion f] (H : P g) :
+    P (f ≫ g) := by
+  rw [eq_affineLocally P, affineLocally_iff_affineOpens_le] at H ⊢
+  intro U V e
+  have : IsIso (f.appLE (f ''ᵁ V) V.1 (f.preimage_image_eq _).ge) :=
+    inferInstanceAs (IsIso (f.app _ ≫
+      X.presheaf.map (eqToHom (f.preimage_image_eq _).symm).op))
+  rw [← Scheme.appLE_comp_appLE _ _ _ (f ''ᵁ V) V.1
+    (Set.image_subset_iff.mpr e) (f.preimage_image_eq _).ge,
+    (isLocal_ringHomProperty P).respectsIso.cancel_right_isIso]
+  exact H _ ⟨_, V.2.image_of_isOpenImmersion _⟩ _
+
+variable {P f}
+
+lemma iff_appLE : P f ↔ ∀ (U : Y.affineOpens) (V : X.affineOpens) (e), Q (f.appLE U V e) := by
+  rw [eq_affineLocally P, affineLocally_iff_affineOpens_le]
+
+theorem of_openCover [IsAffine Y]
+    (𝒰 : X.OpenCover) [∀ i, IsAffine (𝒰.obj i)] (H : ∀ i, Q ((𝒰.map i ≫ f).app ⊤)) :
+    P f := by
```

Comment:
Maybe the name should indicate that the cover is of the source, e.g. `of_openCover_source`? Maybe we can make the convention that `openCover` alway relates to the target and `openCover_source` to the source.